src/plugins/lexer/mod.ts

```diff
@@ -0,0 +1,390 @@
+// import some types/const enums to keep source tidy
+import { Mode, Token, Trace } from "./types/mod.ts";
+
+/*
+note:
+  lexer uses zero-based numbering for tracking
+  its position and the position of tokens.
+*/
+export class Lexer {
+  source: string = "";
+  // lexer position on file
+  position: number = -1;
+  // virtual position; used for tracking token positions.
+  virtual_position: number = -1;
+  // keeps track of elementary debug info useful for parser
+  trace: Trace = {
+    line: 1,
+    column: [0, 0],
+  };
+  // stores current built token
+  token: string = "";
+  // stores current lexical tokens
+  tokens: Token[] = [];
+  // stores mode lexer is interpreting in
+  mode: Mode = Mode.Token;
+
+  // optional initiator for .mtl source
+  constructor(input?: string) {
+    if (input !== undefined) {
+      this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
+    }
+  }
+
+  // loads new .mtl source, avoiding reinitializing class
+  load(input: string) {
+    this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
```

Comment:
Responding to 1: not sure what is meant by "end of a line is in the middle of a string" 😅. Internal eol tokens are used for keeping track of virtual line placement and possible use for future feature additions in MTL. Could you rephrase?

As a general thought, I feel we should ditch multi-line strings in replacement for chain-able single lines. Escape sequences will remain such as `\n` and friends, which will not inhibit loading external files into unsafe blocks, but multi-line strings are *very* error prone and splitting them up could help with both string layout and error handling.

**Examples**
_note: indentation used is four spaces_
With multi-line:
```
const my_sentence: string {
    "The quick brown fox
jumps over the lazy dog"
}
```
with single line:
```
const my_sentence: string {
    "The quick brown fox"
    "jumps over the lazy dog"
}
# or
const my_sentence: string {
    "The quick brown fox\njumps over the lazy dog"
}
```
Output:
```
The quick brown fox
jumps over the lazy dog
```

Continuing, we can also experience weird formatting issues when trying to tidy multi-line strings.
With multi-line:
```
const my_sentence: string {
    "The quick brown fox
     jumps over the lazy dog"
}

```
Output:
```
The quick brown fox
     jumps over the lazy dog
```

**On the topic of error handling**, with single line the lexer is capable of lexing on and finding other concurring errors without searching for a closing quotation on unclosed strings.

With multi-line:
```
# in this example, the lexer is stuck in string mode
# and unable to detect token errors, with each character
# character being loaded into the string.

cost unclosed_string: string {
    "I'm a unclosed
multine string...
}

# -- continued source in file--
```

With single line:
```
cost unclosed_string: string {
    "I'm a unclosed"
    "multiline string...  [lexer catches unclosed string here]
    [lexer marks string as unclosed and continues lexing tokens to discover possible other syntax errors]
    
# -- continued source in file--
```

Reply:
When you initially brought up this idea I really wasn't a fan -- I thought using multi-line was a much better approach that would allow for flexibility within the language. After reading your rationale though, I actually really like this idea. Error handling definitely becomes easier, but even further, I think this adds to our logical syntax design approach... whitespace & spacing becomes much more visible when reading code using this method (+ whitespace was annoying to implement anyway, so fine by me haha).

Before resolving:
- Come up with a method of knowing when someone used a multiline. Right now the second part of a multiline will be considered an ident, which shouldn't be the case. I'm pretty sure once [this](https://github.com/mtl-lang/MTL/pull/1#discussion_r999931136) bug is fixed, it'll also resolve this issue. Example:
```
const x: string { "hey
      there" };
```
```
  { type: "invalid_string", value: "hey", trace: { line: 1, column: [ 18, 21 ] } },
  { type: "eol", value: "*", trace: { line: 1, column: [ -1, -1 ] } },
  { type: "ident", value: 'there"', trace: { line: 2, column: [ 6, 11 ] } },
  ```
  
  - Do some testing with whitespace. I think this *could* have some wonky implications that we aren't thinking of, mainly because it's unconventional, and should be extensively tested to make sure it makes sense and is readable in an actual codebase.

Reply:
After thinking about this some more today, I have some additional thoughts.

How do we plan to actually parse this syntax? The lexer will produce two string tokens, even though a variable only has one string value, and then what -- do we expect the compiler to append them during codegen? That doesn't make much sense to me.

In your example we had the following src:
```
const my_sentence: string {
    "The quick brown fox"
    "jumps over the lazy dog"
}
```
Which resulted in an output of: 
```
The quick brown fox
jumps over the lazy dog
```
and was equivalent to `The quick brown fox\njumps over the lazy dog`

So this leads to three important questions: 1. When are these strings appended to each other, 2. Does this break Lydo conventions, and 3. Does this add unnecessary complexity for the purpose of being cute?

For the POV of the programmer, our variable `my_sentence` should contain a single string: `The quick brown fox\n jumps over the lazy dog`, yet by reading the code it seems that there are two strings, almost like an array of two strings, which isn't an added confusion we want.

Even if that confusion doesn't exist though, Lydo's current structure is to return the very last element in a set of braces. With this in mind, aren't we breaking the convention to return multiple elements (even if they're appended to end up as one element)? What happens if I have some code like this:
```
const my_word: string { "hi" }
const my_sentence: string {
    "my word is:"
    @my_word
    "this ^"
}
```
With the proposed method, we expect `my word is:\nhi\nthis^` -- this is super hard to determine by just reading the src. With our current return structure for vars, it's expected that you can read the very last line and see what the value of `my_sentence` will be -- here that isn't the case, and that adds a ton of ambiguity. 

When you start combining syntax into one data type things get kinda iffy. And this goes back to the first question -- when are these actually appended? It's expected that `my_sentence` is only one string, which means our sets of quotes are combined at some point and joined with a `\n`... if this occurs at codegen time, I feel like it could introduce bugs or generally be very confusing for the end developer (and us while building this) to write code. Not to mention -- when has this syntax, or anything like it, been used by another language? If you're learning the language for the first time, It would be VERY confusing to imagine two separate strings being combined, and joined with a linebreak, simply because they're next to each other in the code. I imagine someone learning the language saying to themself: "So it's two strings with four quotes, but actually it's one string made up of two appended strings which for some reason are also joined with a linebreak, and it's returned as a single string data type" -- yikes. 

I still like the concept, but doubtful about how this would be executed. This was the one comment on this PR that I couldn't get out of my head today. The more I thought about it, the more I remembered Ryan Dahl's quote from his Node Regrets talk: "What I've come to learn now that I'm ageing is that whenever you're designing a program there are things that you think might be cute to add in. You always regret those."

Reply:
Responding to the thoughts in order:

I definitely should have been more detailed with explanation of how it would be handled, as I can see how confusion can come about.

Based on example:
```
const my_sentence: string {
    "The quick brown fox"
    "jumps over the lazy dog"
}
```
In the above source, the parser will output two string tokens, which isn't much of an issue as it's very clear to the parser what is implied with strings side by side. Even with explicit operators, the source is quite understandable:

```
const my_sentence: string {
    "The quick brown fox" +
    "jumps over the lazy dog"
}
```
As for variables producing two string values.. MTL doesn't have the concept of a variable by design, only collections with everything grouped together. *At code generation* the above code would be transformed into a variable, but that has nothing to do with MTL itself. Appending would be a parser step, transforming the strings into:

`The quick brown fox[&newline]jumps over the lazy dog`
(note: newline escape is not used, as it doesn't exist in all programming languages and is based off C dialect for parser/ast replacement. As for the parser insert, that's to be determined during design.)

As to answer the numbered questions:
`When are these strings appended to each other?`:
Strings are appended in parser/AST tree generation stage. Definitely not something left to later stages, as code generation should not need to focus on details about MTL.

`Does this break Lydo conventions?`:
MTL conventions are not formalized as of yet, and there's already contradicting implementations for the "return last line approach" we discussed, visibly in collections:
```
const sample: collection {
    @my_variable_1,
    @my_variable_2,
    @my_variable_3,
}
```

`3. Does this add unnecessary complexity for the purpose of being cute?`:
Definitely an interesting question, and what I disagreed with the most. With your approach of multi-line strings and the issue of formatting below:
```
const my_sentence: string {
    "The quick brown fox
     jumps over the lazy dog"
}
```
The developer ends up needlessly writing code to remove front line spacing or just ends up using single line regardless:
```
# no formatting issue
const my_sentence: string {
    "The quick brown fox" +
     "jumps over the lazy dog"
}
```
The single line approach without needing to infer `+` after each string to reach desired multi-line strings formatting is similar to just using `+` after each string

Going back to the discussion of Lydo conventions, would this not be invalid?:
```
const my_number: number {
    1
    + 1
    - 3 * 6
    -17 + 84
}
```

At the end of the day, it's removing a detrimental issue that could mess up the desired error reporting, avoiding unwanted string formatting issues, adding a simple sugar of not needing a `+` after each string, while optimizing the output through the parser in a simple step.

`yet by reading the code it seems that there are two strings, almost like an array of two strings, which isn't an added confusion we want.`
Array is again language specific and not in all languages. MTL's goal is to approach issues uniquely, and break from existing language design if it means improvement in what we see fit for developer experience.

Answering the next concern:
`What happens if I have some code like this?`
```
const my_word: string { "hi" }
const my_sentence: string {
    "my word is:"
    @my_word
    "this ^"
}
```
That is invalid source, as the variable should be placed in a string:
```

const my_word: string { "hi" }
const my_sentence: string {
    "my word is:"
    "@my_word"
    "this ^"
}
```
Variables inside a string was a discussion topic which was agreed on, and deviation from that could generate confusing syntax.
The above proper example would compile to such in JS:
```js
const my_word = "hi";

const my_string = `my word is:\n${my_word}\nthis ^`;
// or alternatively
const my_string = "my word is:" + "\n" + ${my_word} + "\nthis ^";
```

For the concern of appending happening during codegen, that is 100% not the case. Nothing pre-generation should be *ever* be pushed onto code generation to handle. Lexer interprets tokens, parser analyzes tokens and builds a code execution tree, IR code (if added for code generation optimizations later on) compiles parser tree (rinse and repeat lexer/parser steps for IR) and finally code generation takes a look at final syntax tree and simply translates to target language.

`when has this syntax, or anything like it, been used by another language?`
Why should it have to? There always has to be a first.

```
"So it's two strings with four quotes, but actually it's one string made up of two appended strings which for some reason are also joined with a linebreak, and it's returned as a single string data type" -- yikes.
```
Documentation exists to explain these types of things. It's the same as looking into Rusts abstractions and thinking it's terrible design of abstracting allocations without reading into docs. Like anything, there's a learning curve, but I feel what's intended is very visible. If not, there's a documentation section.

Ending off, there's definitely a divide between "cute" and functional. Error handling is definitely the most important point of the single line approach.

If still disagreed upon, I'm willing to go the multi-line approach that you have in mind, but really do think it will sting long term.

-- response is personal thoughts --

Reply:
Some of these points are really good, but I still think this needs further discussion.

Just to make sure my intentions are clear -- I actually don't have a preference on method. These comments:
> the multi-line approach that you have in mind [...] With your approach of multi-line strings 

are a little misrepresentative of my intention. I'm more so concerned about the flaws with *both* of the implementations we've thought of thus far; I don't think either is a great option. Because of this, I want to make sure we discuss this thouroughly and get it right. 

We've outlined a bunch of pros and cons, and I think this thread has a good collection of thoughts. Let's have a call and discuss this in further detail. Afterward, we can write a summary in this thread of our decision & its rationale.